File: legacy/engine/activities.py
"""Temporal activities: every DB / storage / client touch of the engine.

Engine node bodies execute here (sync, thread-pooled). Human tasks are
standalone Temporal workflows started here with hard id-dedupe
('node-{instance}-{engagement}-{memo_key}', conflict policy USE_EXISTING).
"""

from __future__ import annotations

import asyncio
import inspect
from typing import Any

from temporalio import activity
from temporalio.client import Client
from temporalio.common import WorkflowIDConflictPolicy
from temporalio.exceptions import ApplicationError

from engine import db as dbm
from engine.canonical import canonical_bytes
from engine.context import ArtifactHandle
from engine.registry import REGISTRY, HumanTask


def _decode(value: Any, loader) -> Any:
    if isinstance(value, dict):
        if "__artifact__" in value and len(value) == 1:
            return ArtifactHandle(value["__artifact__"], loader=loader)
        return {k: _decode(v, loader) for k, v in value.items()}
    if isinstance(value, list):
        return [_decode(v, loader) for v in value]
    return value


def _encode_payload_value(value: Any) -> Any:
    if isinstance(value, ArtifactHandle):
        return {"__artifact__": dict(value.ref)}
    if isinstance(value, dict):
        return {k: _encode_payload_value(v) for k, v in value.items()}
    if isinstance(value, list):
        return [_encode_payload_value(v) for v in value]
    return value


def _to_output_bytes(result: Any) -> tuple[bytes, str]:
    if isinstance(result, bytes):
        return result, "application/octet-stream"
    if isinstance(result, str):
        return result.encode("utf-8"), "text/plain"
    return canonical_bytes(result), "application/json"


class GraphflowActivities:
    def __init__(self, db_path: str, storage_root: str, client: Client,
                 task_queue: str, instance: str):
        self.db_path = db_path
        self.storage_root = storage_root
        self.client = client
        self.task_queue = task_queue
        self.instance = instance

    def _conn(self):
        return dbm.connect(self.db_path)

    def _loader(self, conn):
        return lambda artifact_id: dbm.read_artifact_payload(
            conn, self.storage_root, artifact_id
        )

    # ---------- lookups / attach ----------

    @activity.defn(name="memo_lookup")
    def memo_lookup(self, engagement_id: int, memo_key: str) -> dict | None:
        conn = self._conn()
        try:
            return dbm.memo_lookup(conn, engagement_id, memo_key)
        finally:
            conn.close()

    @activity.defn(name="attach_artifact")
    def attach_artifact(self, workflow_run_id: int, artifact_id: int) -> None:
        conn = self._conn()
        try:
            dbm.attach(conn, workflow_run_id, artifact_id, source="engine", added_by="engine")
        finally:
            conn.close()

    # ---------- engine node execution ----------

    @activity.defn(name="run_engine_node")
    def run_engine_node(self, req: dict) -> dict:
        conn = self._conn()
        try:
            # Idempotency on activity retry: the question may already be answered.
            existing = dbm.memo_lookup(conn, req["engagement_id"], req["memo_key"])
            if existing is not None:
                return {"ref": existing, "fresh": False}

            nd = REGISTRY.node_for_workflow(req["workflow_id"], req["node_id"])
            kwargs = _decode(req["args_transport"], self._loader(conn))
            # Node-body exceptions propagate as ordinary (retryable) activity
            # failures — the 5-attempt policy absorbs transients. Authors raise
            # ApplicationError(type="NodeError") for permanently-bad inputs.
            result = nd.fn(**kwargs)
            if inspect.iscoroutine(result):
                result = asyncio.run(result)

            try:
                payload, media_type = _to_output_bytes(result)
            except ValueError as exc:
                raise ApplicationError(
                    f"node {nd.node_id} produced a non-canonical payload: {exc}",
                    type="NodeError",
                    non_retryable=True,
                ) from exc
            info = activity.info()
            temporal_id = f"{info.workflow_id}/{info.workflow_run_id}/{info.activity_id}"
            ref, fresh = dbm.record_completion(
                conn,
                self.storage_root,
                engagement_id=req["engagement_id"],
                workflow_run_id=req["workflow_run_id"],
                workflow_id=req["workflow_id"],
                node_id=req["node_id"],
                code_hash=nd.code_hash,
                memo_key=req["memo_key"],
                output_kind=nd.output_kind,
                payload=payload,
                media_type=media_type,
                created_by="engine",
                temporal_id=temporal_id,
                input_artifact_ids=req["input_artifact_ids"],
            )
            return {"ref": ref, "fresh": fresh}
        finally:
            conn.close()

    # ---------- human tasks ----------

    @activity.defn(name="ensure_human_task")
    async def ensure_human_task(self, req: dict) -> str:
        """Build the HumanTask (question) and start-or-attach the task workflow.
        Hard dedupe by workflow id; the task workflow itself re-checks the memo
        as its first step, closing the start-after-completion race."""
        conn = self._conn()
        try:
            nd = REGISTRY.node_for_workflow(req["workflow_id"], req["node_id"])
            kwargs = _decode(req["args_transport"], self._loader(conn))
            try:
                ht = nd.fn(**kwargs)
            except Exception as exc:
                # A crashing question-builder is deterministic: fail the run
                # visibly instead of retrying forever.
                raise ApplicationError(
                    f"human node {nd.node_id} failed building its task: {exc}",
                    type="NodeError",
                    non_retryable=True,
                ) from exc
            if not isinstance(ht, HumanTask):
                raise ApplicationError(
                    f"human node {nd.node_id} must return a HumanTask",
                    non_retryable=True,
                )
            task_input = {
                "engagement_id": req["engagement_id"],
                "workflow_id": req["workflow_id"],
                "node_id": req["node_id"],
                "code_hash": nd.code_hash,
                "memo_key": req["memo_key"],
                "output_kind": nd.output_kind,
                "display_name": nd.display_name,
                "instructions": ht.instructions,
                "payload": _encode_payload_value(ht.payload),
                "result_required_keys": ht.result_required_keys,
                "requested_by_workflow_run": req["workflow_run_id"],
                "input_artifact_ids": req["input_artifact_ids"],
            }
        finally:
            conn.close()

        task_wf_id = f"node-{self.instance}-{req['engagement_id']}-{req['memo_key']}"
        # No exception handling on purpose: with USE_EXISTING (+ default
        # allow-duplicate reuse) an existing or completed task never raises —
        # a start after completion spawns a run that self-completes via its
        # first-step memo check. Anything that DOES raise here is a genuine
        # failure and must surface so the activity retries instead of leaving
        # the requester polling for a task that was never created.
        await self.client.start_workflow(
            "GraphflowHumanTask",
            task_input,
            id=task_wf_id,
            task_queue=self.task_queue,
            id_conflict_policy=WorkflowIDConflictPolicy.USE_EXISTING,
        )
        return task_wf_id

    @activity.defn(name="record_human_completion")
    def record_human_completion(self, task_input: dict, result: dict, reviewer: str) -> dict:
        conn = self._conn()
        try:
            try:
                payload = canonical_bytes(result)
            except ValueError as exc:
                # Deterministic: retrying cannot help. The submit validator
                # rejects this before it reaches us; this is the belt-and-braces.
                raise ApplicationError(
                    f"submission is not canonicalizable: {exc}", non_retryable=True
                ) from exc
            info = activity.info()
            temporal_id = f"{info.workflow_id}/{info.workflow_run_id}"
            ref, _fresh = dbm.record_completion(
                conn,
                self.storage_root,
                engagement_id=task_input["engagement_id"],
                workflow_run_id=task_input.get("requested_by_workflow_run"),
                workflow_id=task_input["workflow_id"],
                node_id=task_input["node_id"],
                code_hash=task_input["code_hash"],
                memo_key=task_input["memo_key"],
                output_kind=task_input["output_kind"],
                payload=payload,
                media_type="application/json",
                created_by=reviewer,
                temporal_id=temporal_id,
                input_artifact_ids=task_input["input_artifact_ids"],
            )
            return ref
        finally:
            conn.close()

    def all(self) -> list:
        return [
            self.memo_lookup,
            self.attach_artifact,
            self.run_engine_node,
            self.ensure_human_task,
            self.record_human_completion,
        ]

File: backend/src/temporal/Activities.ts
import { activityInfo } from '@temporalio/activity';
import type { Client } from '@temporalio/client';
import { ApplicationFailure } from '@temporalio/common';
import { ArtifactHandle, type PayloadLoader } from '../domain/artifact/ArtifactHandle.js';
import type { ArtifactRef } from '../domain/artifact/ArtifactRef.js';
import { canonicalBytes } from '../domain/canonical/Canonical.js';
import type { JsonValue } from '../domain/json/JsonValue.js';
import type { HumanTask, NodeArgValue, NodeResult, Registry } from '../domain/registry/Registry.js';
import { attach, connect, memoLookup, readArtifactPayload, recordCompletion } from '../infrastructure/db/Db.js';
import { errorMessage } from '../shared/errors/Errors.js';
import type { NodeRequest, TransportValue } from './Context.js';
import { HUMAN_TASK_WORKFLOW_TYPE, humanTaskWorkflowId } from './Ids.js';
import type { TaskInput } from './Workflows.js';

// Node-side: every DB / storage / client touch. Engine node bodies execute here. Human tasks are
// standalone Temporal workflows started here with hard id-dedupe
// ('node-{instance}-{engagement}-{memo_key}', conflict policy USE_EXISTING).

export interface ActivityDeps {
  dbPath: string;
  storageRoot: string;
  client: Client;
  taskQueue: string;
  instance: string;
  registry: Registry;
}

const textEncoder = new TextEncoder();

// A dict is an artifact marker ONLY if __artifact__ is its sole key.
const isArtifactMarker = (value: object): value is { __artifact__: ArtifactRef } => {
  const keys = Object.keys(value);
  return keys.length === 1 && keys[0] === '__artifact__';
};

function decodeTransport(value: TransportValue, loader: PayloadLoader): NodeArgValue {
  if (value === null || typeof value !== 'object') {
    return value;
  }
  if (Array.isArray(value)) {
    return value.map((item) => decodeTransport(item, loader));
  }
  if (isArtifactMarker(value)) {
    return new ArtifactHandle(value.__artifact__, loader);
  }
  const out: { [key: string]: NodeArgValue } = {};
  for (const [k, v] of Object.entries(value)) {
    out[k] = decodeTransport(v, loader);
  }
  return out;
}

function decodeArgs(transport: TransportValue, loader: PayloadLoader): Record<string, NodeArgValue> {
  const decoded = decodeTransport(transport, loader);
  if (decoded === null || typeof decoded !== 'object' || Array.isArray(decoded) || decoded instanceof ArtifactHandle) {
    throw ApplicationFailure.nonRetryable('args_transport must be an object of named arguments');
  }
  return decoded;
}

// Inverse of decodeTransport, used for HumanTask.payload.
function encodePayloadValue(value: NodeArgValue): TransportValue {
  if (value instanceof ArtifactHandle) {
    return { __artifact__: { ...value.ref } };
  }
  if (Array.isArray(value)) {
    return value.map(encodePayloadValue);
  }
  if (value !== null && typeof value === 'object') {
    const out: { [key: string]: TransportValue } = {};
    for (const [k, v] of Object.entries(value)) {
      out[k] = encodePayloadValue(v);
    }
    return out;
  }
  return value;
}

// Node return-value contract: bytes -> octet-stream; string -> UTF-8 text/plain; anything else
// must be canonical-JSON-safe (canonicalBytes throws on floats etc.).
function toOutputBytes(result: NodeResult): { payload: Uint8Array; mediaType: string } {
  if (result instanceof Uint8Array) {
    return { payload: result, mediaType: 'application/octet-stream' };
  }
  if (typeof result === 'string') {
    return { payload: textEncoder.encode(result), mediaType: 'text/plain' };
  }
  return { payload: canonicalBytes(result as JsonValue), mediaType: 'application/json' };
}

// Structural type guard: HumanTask crosses the activity boundary as plain data, never as a class
// instance, so a shape check (not instanceof) is the only way to recognize one.
function isHumanTask(value: NodeResult): value is HumanTask {
  if (value === null || typeof value !== 'object' || Array.isArray(value) || value instanceof Uint8Array) {
    return false;
  }
  return (
    typeof value.instructions === 'string' &&
    Array.isArray(value.resultRequiredKeys) &&
    value.payload !== null &&
    typeof value.payload === 'object' &&
    !Array.isArray(value.payload)
  );
}

function schedulingWorkflow(): { workflowId: string; runId: string } {
  const execution = activityInfo().workflowExecution;
  if (execution === undefined) {
    throw ApplicationFailure.nonRetryable('activity was not scheduled by a workflow');
  }
  return execution;
}

// Build the HumanTask (question) from the node body; the DB connection is closed before the
// caller talks to Temporal.
async function buildTaskInput(deps: ActivityDeps, req: NodeRequest): Promise<TaskInput> {
  const conn = connect(deps.dbPath);
  try {
    const registered = deps.registry.nodeForWorkflow(req.workflow_id, req.node_id);
    const loader: PayloadLoader = async (artifactId) => readArtifactPayload(conn, deps.storageRoot, artifactId);
    const args = decodeArgs(req.args_transport, loader);
    let task: NodeResult;
    try {
      task = await registered.run(args);
    } catch (e) {
      // A crashing question-builder is deterministic: fail the run visibly instead of retrying forever.
      throw ApplicationFailure.create({
        message: `human node ${registered.nodeId} failed building its task: ${errorMessage(e)}`,
        type: 'NodeError',
        nonRetryable: true,
      });
    }
    if (!isHumanTask(task)) {
      throw ApplicationFailure.nonRetryable(`human node ${registered.nodeId} must return a HumanTask`);
    }
    return {
      engagement_id: req.engagement_id,
      workflow_id: req.workflow_id,
      node_id: req.node_id,
      memo_key: req.memo_key,
      output_nodeparamslot: registered.outputNodeparamslot,
      display_name: registered.displayName,
      instructions: task.instructions,
      payload: encodePayloadValue(task.payload),
      result_required_keys: task.resultRequiredKeys,
      requested_by_workflow_run: req.workflow_run_id,
      input_artifact_ids: req.input_artifact_ids,
    };
  } finally {
    conn.close();
  }
}

// The returned object IS the worker registration payload; keys are the wire activity names.
export function createActivities(deps: ActivityDeps) {
  return {
    async memo_lookup(engagementId: number, memoKey: string): Promise<ArtifactRef | null> {
      const conn = connect(deps.dbPath);
      try {
        return memoLookup(conn, engagementId, memoKey);
      } finally {
        conn.close();
      }
    },

    async attach_artifact(workflowRunId: number, artifactId: number): Promise<void> {
      const conn = connect(deps.dbPath);
      try {
        attach(conn, workflowRunId, artifactId, { source: 'engine', createdBy: 'engine' });
      } finally {
        conn.close();
      }
    },

    async run_engine_node(req: NodeRequest): Promise<{ ref: ArtifactRef; fresh: boolean }> {
      const conn = connect(deps.dbPath);
      try {
        // Idempotency on activity retry: the completion may already be filed (no attach here —
        // Ctx attaches on non-fresh).
        const existing = memoLookup(conn, req.engagement_id, req.memo_key);
        if (existing !== null) {
          return { ref: existing, fresh: false };
        }
        const registered = deps.registry.nodeForWorkflow(req.workflow_id, req.node_id);
        const loader: PayloadLoader = async (artifactId) => readArtifactPayload(conn, deps.storageRoot, artifactId);
        const args = decodeArgs(req.args_transport, loader);
        // Node-body exceptions propagate as ordinary (retryable) activity failures — the
        // 5-attempt policy absorbs transients; authors throw type 'NodeError' for bad inputs.
        const result = await registered.run(args);
        let output: { payload: Uint8Array; mediaType: string };
        try {
          output = toOutputBytes(result);
        } catch (e) {
          throw ApplicationFailure.create({
            message: `node ${registered.nodeId} produced a non-canonical payload: ${errorMessage(e)}`,
            type: 'NodeError',
            nonRetryable: true,
          });
        }
        const execution = schedulingWorkflow();
        const temporalId = `${execution.workflowId}/${execution.runId}/${activityInfo().activityId}`;
        return recordCompletion(conn, deps.storageRoot, {
          engagementId: req.engagement_id,
          workflowRunId: req.workflow_run_id,
          workflowId: req.workflow_id,
          nodeId: req.node_id,
          memoKey: req.memo_key,
          outputNodeparamslot: registered.outputNodeparamslot,
          payload: output.payload,
          mediaType: output.mediaType,
          createdBy: 'engine',
          temporalId,
          inputArtifactIds: req.input_artifact_ids,
        });
      } finally {
        conn.close();
      }
    },

    async ensure_human_task(req: NodeRequest): Promise<string> {
      const taskInput = await buildTaskInput(deps, req);
      const taskWfId = humanTaskWorkflowId(deps.instance, req.engagement_id, req.memo_key);
      // No exception handling on purpose: with USE_EXISTING an existing or completed task never
      // raises — a start after completion spawns a run that self-completes via its first-step
      // memo check. Anything that DOES raise is a genuine failure and must surface so the
      // activity retries instead of leaving the requester polling for a task never created.
      await deps.client.workflow.start(HUMAN_TASK_WORKFLOW_TYPE, {
        args: [taskInput],
        workflowId: taskWfId,
        taskQueue: deps.taskQueue,
        workflowIdConflictPolicy: 'USE_EXISTING',
      });
      return taskWfId;
    },

    async record_human_completion(
      taskInput: TaskInput,
      result: Record<string, JsonValue>,
      reviewer: string
    ): Promise<ArtifactRef> {
      let payload: Uint8Array;
      try {
        // Deterministic — retrying cannot help; the submit validator rejects this earlier
        // (belt-and-braces).
        payload = canonicalBytes(result);
      } catch (e) {
        throw ApplicationFailure.nonRetryable(`submission is not canonicalizable: ${errorMessage(e)}`);
      }
      const execution = schedulingWorkflow();
      // Two segments only, no activity_id (unlike run_engine_node).
      const temporalId = `${execution.workflowId}/${execution.runId}`;
      const conn = connect(deps.dbPath);
      try {
        const completion = recordCompletion(conn, deps.storageRoot, {
          engagementId: taskInput.engagement_id,
          workflowRunId: taskInput.requested_by_workflow_run,
          workflowId: taskInput.workflow_id,
          nodeId: taskInput.node_id,
          memoKey: taskInput.memo_key,
          outputNodeparamslot: taskInput.output_nodeparamslot,
          payload,
          mediaType: 'application/json',
          createdBy: reviewer,
          temporalId,
          inputArtifactIds: taskInput.input_artifact_ids,
        });
        return completion.ref;
      } finally {
        conn.close();
      }
    },
  };
}

export type Acts = ReturnType<typeof createActivities>;
